src/assets/wise5/classroomMonitor/dataExport/strategies/AbstractComponentDataExportStrategy.ts

```diff
@@ -0,0 +1,393 @@
+    columnNameToNumber: any,
+    nodeId: string,
+    componentId: string,
+    rowCounter: number
+  ): string[] {
+    const rows = [];
+    const userInfo = this.configService.getUserInfoByWorkgroupId(workgroupId);
+    const extractedUserIDsAndStudentNames = this.extractUserIDsAndStudentNames(userInfo.users);
+
+    // A mapping from component to component revision counter. The key will be
+    // {{nodeId}}_{{componentId}} and the value will be a number.
+    const componentRevisionCounter = {};
+    const componentStates = this.teacherDataService.getComponentStatesByWorkgroupIdAndComponentId(
+      workgroupId,
+      componentId
+    );
+    for (let c = 0; c < componentStates.length; c++) {
+      const componentState = componentStates[c];
+      if (this.shouldExportRow(componentState, c, componentStates.length)) {
+        const row = this.generateComponentWorkRow(
+          component,
+          columnNames,
+          columnNameToNumber,
+          rowCounter,
+          workgroupId,
+          extractedUserIDsAndStudentNames['userId1'],
+          extractedUserIDsAndStudentNames['userId2'],
+          extractedUserIDsAndStudentNames['userId3'],
+          extractedUserIDsAndStudentNames['studentName1'],
+          extractedUserIDsAndStudentNames['studentName2'],
+          extractedUserIDsAndStudentNames['studentName3'],
+          userInfo.periodName,
+          componentRevisionCounter,
+          componentState
+        );
+        rows.push(row);
+        rowCounter++;
+      } else {
+        // We do not want to add this component state as a row in the export but we still want to
+        // increment the revision counter.
+        this.incrementRevisionCounter(componentRevisionCounter, nodeId, componentId);
+      }
+    }
+    return rows;
+  }
+
+  /**
+   * @param users An array of user objects. Each user object contains an id and name.
+   * @returns {object} An object that contains key/value pairs. The key is userIdX
+   * or studentNameX where X is an integer. The values are the corresponding actual
+   * values of user id and student name.
+   */
+  private extractUserIDsAndStudentNames(users: any[]): any {
+    const extractedUserIDsAndStudentNames = {};
+    for (let u = 0; u < users.length; u++) {
+      const user = users[u];
+      extractedUserIDsAndStudentNames['userId' + (u + 1)] = user.id;
+      if (this.canViewStudentNames) {
+        extractedUserIDsAndStudentNames['studentName' + (u + 1)] = user.name;
+      }
+    }
+    return extractedUserIDsAndStudentNames;
+  }
+
+  private shouldExportRow(
+    componentState: any,
+    componentStateIndex: number,
+    numComponentStates: number
+  ): boolean {
+    let exportRow = true;
+    if (this.includeOnlySubmits && !componentState.isSubmit) {
+      exportRow = false;
+    } else if (
+      this.workSelectionType === 'exportLatestWork' &&
+      componentStateIndex != numComponentStates - 1
+    ) {
+      exportRow = false;
+    }
+    return exportRow;
```

Comment:
This block is a bit hard to read. Can this be re-written without the exportRow variable, and combine the two false cases?
```
if ((this.includeOnlySubmits && !componentState.isSubmit) ||
    (this.workSelectionType === 'exportLatestWork' && componentStateIndex != numComponentStates - 1)) {
  return false;
}
return true;
```

It would be even better if we could put the ```this.includeOnlySubmits && !componentState.isSubmit``` and ```this.workSelectionType === 'exportLatestWork' && componentStateIndex != numComponentStates - 1``` blocks into their own function with a descriptive function name? Then this function would be one line:
```return !(function1() || function2())```